src/arch/wasm/CodeGen.zig

```diff
@@ -5821,6 +5821,99 @@ fn airPopcount(func: *CodeGen, inst: Air.Inst.Index) InnerError!void {
     func.finishAir(inst, result, &.{ty_op.operand});
 }
 
+fn airBitReverse(func: *CodeGen, inst: Air.Inst.Index) InnerError!void {
+    const mod = func.bin_file.base.comp.module.?;
+    const ty_op = func.air.instructions.items(.data)[@intFromEnum(inst)].ty_op;
+
+    const operand = try func.resolveInst(ty_op.operand);
+    const ty = func.typeOf(ty_op.operand);
+
+    if (ty.zigTypeTag(mod) == .Vector) {
+        return func.fail("TODO: Implement @bitReverse for vectors", .{});
+    }
+
+    const int_info = ty.intInfo(mod);
+    const bits = int_info.bits;
+    const wasm_bits = toWasmBits(bits) orelse {
+        return func.fail("TODO: Implement @bitReverse for integers with bitsize '{d}'", .{bits});
+    };
+
+    switch (wasm_bits) {
+        32 => {
+            const intrin_ret = try func.callIntrinsic(
+                "__bitreversesi2",
+                &.{.u32_type},
+                Type.u32,
+                &.{operand},
+            );
+            const reversed = if (bits == 32)
+                intrin_ret
+            else
+                try func.binOp(intrin_ret, .{ .imm32 = 32 - bits }, Type.u32, .shr);
+            const result = try reversed.toLocal(func, ty);
+            func.finishAir(inst, result, &.{ty_op.operand});
+        },
+        64 => {
+            const intrin_ret = try func.callIntrinsic(
+                "__bitreversedi2",
+                &.{.u64_type},
+                Type.u64,
+                &.{operand},
+            );
+            const reversed = if (bits == 64)
+                intrin_ret
+            else
+                try func.binOp(intrin_ret, .{ .imm64 = 64 - bits }, Type.u64, .shr);
+            const result = try reversed.toLocal(func, ty);
+            func.finishAir(inst, result, &.{ty_op.operand});
+        },
+        128 => {
+            const result = try func.allocStack(ty);
+
+            try func.emitWValue(result);
+            const first_half = try func.load(operand, Type.u64, 8);
+            const intrin_ret_first = try func.callIntrinsic(
```

Comment:
Might be worth adding a small comment here mentioning we can replace this logic with a call to `__bitreverseti2` when the Wasm backend supports 128-bit integers on the C-ABI.

Reply:
__bitreverseti2 was not used here for different reasons.

1. When we will support >128 bits integers, this code will change into a for loop calling `__bitreverse` for `u64`, in this case 128 bits will be handled the same as 128 192 256 ... bits integers.

2. I believe that `__bitreverseti2` should be used only if your target architecture supports 128 bit registers (RV128I?) for codegen quality and perf. Also `__bitreverse` for `u64` could potentially be recognized by wasm runtime as bitreverse and lowered better (hand waving, in future we should analyze how chromium/firefox engines lowers our codegen vs llvm vs prefered). A code blob that __bitreverseti2 will generate, unlikely ever to be optimized out.

Reply:
That makes sense, thank you for clarifying.